File: LeetCode/Easy/0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        count = {}
        q = deque([root])

        while q:
            cur = q.popleft()
            count[cur.val] = count.get(cur.val, 0) + 1
            
            if cur.left:
                q.append(cur.left)
            if cur.right:
                q.append(cur.right)

        max_count = max(count.values())

        return [key for key, val in count.items() if val == max_count]

File: LeetCode/Easy/0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        map<int, int> count;
        TreeNode* cur;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            cur = q.front();
            q.pop();
            count[cur->val]++;

            if(cur->left)
                q.push(cur->left);
            if(cur->right)
                q.push(cur->right);
        }

        int max_cnt = 0;
        for(auto &pair : count){
            max_cnt = max(max_cnt, pair.second);
        }

        vector<int> ans;
        for(auto &pair : count){
            if(pair.second == max_cnt)
                ans.push_back(pair.first);
        }
        return ans;
    }
};